engine/apps/mobile_app/tasks.py

```diff
@@ -499,3 +500,192 @@ def conditionally_send_going_oncall_push_notifications_for_schedule(schedule_pk)
+SSR_EARLIEST_NOTIFICATION_OFFSET = datetime.timedelta(weeks=4)
+
+# Once it's time to send out notifications, send them over the course of a week.
+# This is because users can be in multiple timezones / have different working hours configured,
+# so we can't just send all notifications at once, but need to wait for the users to be in their working hours.
+# Once a notification is sent to a user, they won't be notified again for the same shift swap request for a week.
+# After a week, the shift swap request won't be in the notification window anymore (see _get_shift_swap_requests_to_notify).
+SSR_NOTIFICATION_WINDOW = datetime.timedelta(weeks=1)
+
+
+@shared_dedicated_queue_retry_task()
+def notify_shift_swap_requests() -> None:
+    """
+    A periodic task that notifies users about shift swap requests.
+    """
+
+    if not settings.FEATURE_SHIFT_SWAPS_ENABLED:
+        return
+
+    for shift_swap_request in _get_shift_swap_requests_to_notify(timezone.now()):
+        notify_shift_swap_request.delay(shift_swap_request.pk)
+
+
+def _get_shift_swap_requests_to_notify(now: datetime.datetime) -> list[ShiftSwapRequest]:
+    """
+    Returns shifts swap requests that are open and are in the notification window.
+    This method can return the same shift swap request multiple times while it's in the notification window,
+    but users are only notified once per shift swap request (see _mark_shift_swap_request_notified_for_user).
+    """
+
+    shift_swap_requests_in_notification_window = []
+    for shift_swap_request in ShiftSwapRequest.objects.filter(benefactor__isnull=True, swap_start__gt=now):
+        notification_window_start = max(
+            shift_swap_request.created_at, shift_swap_request.swap_start - SSR_EARLIEST_NOTIFICATION_OFFSET
+        )
+        notification_window_end = min(notification_window_start + SSR_NOTIFICATION_WINDOW, shift_swap_request.swap_end)
+
+        if notification_window_start <= now <= notification_window_end:
+            shift_swap_requests_in_notification_window.append(shift_swap_request)
+
+    return shift_swap_requests_in_notification_window
+
+
+@shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True, max_retries=MAX_RETRIES)
+def notify_shift_swap_request(shift_swap_request_pk: int) -> None:
+    """
+    Notify relevant users for an individual shift swap request.
+    """
+    try:
+        shift_swap_request = ShiftSwapRequest.objects.get(pk=shift_swap_request_pk)
+    except ShiftSwapRequest.DoesNotExist:
+        logger.info(f"ShiftSwapRequest {shift_swap_request_pk} does not exist")
+        return
+
+    now = timezone.now()
+    for user in shift_swap_request.possible_benefactors:
+        if _should_notify_user_about_shift_swap_request(shift_swap_request, user, now):
+            notify_user_about_shift_swap_request.delay(shift_swap_request.pk, user.pk)
+            _mark_shift_swap_request_notified_for_user(shift_swap_request, user)
```

Comment:
should this happen at the very end of `notify_user_about_shift_swap_request`? In case something inside there fails we probably don't want to mark the notification as sent. wdyt?

Reply:
I didn't want to do that to make sure the mobile app doesn't spam people. I'd rather not receive any SSR notifications at all than receive more than one (potentially once every 15 minutes if something goes wrong between sending the push notification and marking the notification sent). The `notify_user_about_shift_swap_request` task has a retry policy, so if something goes wrong while sending the push notification (e.g. network issue), the task will probably succeed after a retry. Lmk if this makes sense

Reply:
> task has a retry policy

ahh yes, forgot about this. Makes sense 👍